dev/tasks/r/github.packages.yml

```diff
@@ -279,8 +286,8 @@ jobs:
           {{ macros.github_test_r_src_pkg()|indent(8) }}
 
   upload-binaries:
-    needs: r-packages
-    if: true && !cancelled()
+    # Only upload binaries if all tests pass.

```

Comment:
There's a whole lot of code that munges file names and paths here and then especially on the other side in r_nightly.yml to essentially rebuild the repo directory structure you already have here. What if you just zipped up the whole repo directory and uploaded it as a single artifact? Then in r_nightly.yml you just unzip it and you already have the right directory structure.

Reply:
This would be a nice solution if we had the actual repo structure in this job, which we don't :(. Downloading the artifacts just creates subfolders with the artifact name which are then used to re-name the files to distinguish them. 

I could move the repo-creation code from r_nightly into this workflow but we can't do without it, sorry!

Reply:
But you have `macros.github_setup_local_r_repo`

Reply:
Ah, I see. So a tricky thing to watch out for is the binary package path, which `contrib.url()` will return (https://github.com/ursa-labs/arrow-r-nightly/blob/master/build-binary-and-upload.R#L30). This is different on different R versions and platforms, and unfortunately, CRAN has changed this repo layout periodically--not super frequent, of course, but it seems like with many major releases (annual) there is a change. 

For example, on my machine:

```
> contrib.url("", type="binary")
[1] "/bin/macosx/big-sur-arm64/contrib/4.2"
```

I see the binaries get uploaded as artifacts like this:

```
      - name: Upload binary artifact
        uses: actions/upload-artifact@v3
        with:
          name: r-pkg-{{ '${{ matrix.platform }}-${{ matrix.r_version.r }}' }}
          path: arrow_*
```

I might not be following things correctly, but I don't see how you could end up with that path based on how you're constructing them now. I know we aren't building M1 binaries today, but that's kinda my point: it would be ideal if we didn't have to change the glue code every time we add a new package or when R changes things.

The easiest way to make sure you get the right path is to call contrib.url() from the platform/version you're building on. Maybe instead we create the artifact name based on what contrib.url does, just replacing the path separators with something else, like 

```
> gsub("/", "__", contrib.url("", type="binary"))
[1] "__bin__macosx__big-sur-arm64__contrib__4.2"
```

We can get that into an env var, and then maybe pass it as an `output` so that it can be available in the upload step?

```
 % filename=$(Rscript -e 'cat(gsub("/", "__", contrib.url("", type="binary")))')
 % echo $filename                                                               
__bin__macosx__big-sur-arm64__contrib__4.2
```

Then in r_nightly.yml we gsub back, recursively create the dirs, and that's it. It's a little tortured, but seems less so than trying to reimplement contrib.url?

Reply:
I used `contrib.url` before switching over to the two repo version, good idea! While my current script works and it would not be too hard to add M1 I do agree that we should favor the version that adapts automatically to any CRAN (or R version) changes, so we don't have to manually fix it each time. I'll change this tomorrow :)